src/goto-symex/symex_decl.cpp

```diff
@@ -51,37 +49,23 @@ void goto_symext::symex_decl(statet &state, const symbol_exprt &expr)
   state.rename(ssa.type(), l1_identifier, ns);
   ssa.update_type();
 
-  // in case of pointers, put something into the value set
-  if(ns.follow(expr.type()).id()==ID_pointer)
-  {
-    exprt failed=
-      get_failed_symbol(expr, ns);
-
-    exprt rhs;
-
-    if(failed.is_not_nil())
-      rhs=address_of_exprt(failed, to_pointer_type(expr.type()));
-    else
-      rhs=exprt(ID_invalid);
-
-    state.rename(rhs, ns, goto_symex_statet::L1);
-    state.value_set.assign(ssa, rhs, ns, true, false);
-  }
-
-  // prevent propagation
-  state.propagation.remove(l1_identifier);
-
   // L2 renaming
   // inlining may yield multiple declarations of the same identifier
   // within the same L1 context
   if(state.level2.current_names.find(l1_identifier)==
      state.level2.current_names.end())
     state.level2.current_names[l1_identifier]=std::make_pair(ssa, 0);
-  state.level2.increase_counter(l1_identifier);
-  const bool record_events=state.record_events;
-  state.record_events=false;
-  state.rename(ssa, ns);
-  state.record_events=record_events;
+  init_l1=l1_identifier;
+
+  // skip non-deterministic initialisation if the next instruction
+  // will take care of initialisation (but ensure int x=x; is handled
+  // properly)
+  goto_programt::const_targett next=state.source.pc;
```

Comment:
Seems like a dubious requirement -- for example, this would trip frontends that do
```
int x;
int y;
x = 0;
y = 0;
```
Would it be better to require the front-end to annotate the declaration to indicate whether it has a post-dominating init instruction?

Reply:
Yes, there is definitively code that front ends generate that doesn't benefit from this optimisation. Front ends could provide information, and it may be worth exploring that. For the time being the proposed code will reduce some overhead, without guarantees of catching all cases.